File: code/packages/java/md5-native/src/main/java/com/codingadventures/md5native/Native.java
package com.codingadventures.md5native;

/**
 * JNI bindings to the {@code md5_native_jni} Rust cdylib, which wraps the
 * pure-Rust {@code coding_adventures_md5} crate. Each {@code native} method maps
 * to a {@code Java_com_codingadventures_md5native_Native_*} function.
 */
final class Native {
    static {
        System.loadLibrary("md5_native_jni");
    }

    private Native() {
    }

    static native byte[] nativeDigest(byte[] data);

    static native long nativeHasherNew();

    static native void nativeHasherUpdate(long handle, byte[] data);

    static native byte[] nativeHasherDigest(long handle);

    static native long nativeHasherClone(long handle);

    static native void nativeHasherFree(long handle);
}

File: code/packages/kotlin/md5-native/src/main/kotlin/com/codingadventures/md5native/Native.kt
package com.codingadventures.md5native

/**
 * JNI bindings to the `md5_native_jni` Rust cdylib (shared with
 * `java/md5-native`), which wraps `coding_adventures_md5`. The external methods
 * resolve to the same `Java_com_codingadventures_md5native_Native_*` exports, so
 * Kotlin reuses the existing cdylib with no new Rust crate. MD5 is broken —
 * checksum use only.
 */
internal object Native {
    init {
        System.loadLibrary("md5_native_jni")
    }

    external fun nativeDigest(data: ByteArray): ByteArray
    external fun nativeHasherNew(): Long
    external fun nativeHasherUpdate(handle: Long, data: ByteArray)
    external fun nativeHasherDigest(handle: Long): ByteArray
    external fun nativeHasherClone(handle: Long): Long
    external fun nativeHasherFree(handle: Long)
}
